chore: Bump network-controller to v27 (+ deps)

mcmire opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR bumps `@metamask/network-controller` to 27.0.0.

There are a few notable changes:

- JsonRpcEngine v2 is now being used internally by `createWalletMiddleware`. In this PR we adjust to that and upgrade some of the internal middleware to use it. One of the biggest changes is that request objects cannot have arbitrary properties — that is now available in _context_. You can read more about JsonRpcEngine v2 here: https://github.com/MetaMask/core/blob/main/packages/json-rpc-engine/README.md#usage
- The provider type in `@metamask/eth-json-rpc-provider` changed and is now called `InternalProvider` rather than `SafeEventEmitterProvider`. 
- The connectivity/availability status of an RPC endpoint within NetworkController state is now updated in a more live fashion, i.e., as requests are made instead of being set when it becomes globally selected or when `lookupNetwork` is manually called for the RPC endpoint. This means that anything that is actively relying on network statuses will have more up-to-date data to act on. In particular, this slightly improves the banner shown when the user experiences issues connecting to a network so that if a network becomes degraded or unavailable but then recovers, the banner should disappear for that network more quickly.

Read the changelog for more: https://github.com/MetaMask/core/blob/main/packages/network-controller/CHANGELOG.md

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37966?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Improve responsiveness of network connection banner to connection issues

## **Related issues**

https://consensyssoftware.atlassian.net/browse/WPC-99

## **Manual testing steps**

1. Check out this branch, run `yarn`, run `yarn start`.
2. Load MetaMask, go through onboarding if necessary.
3. Start Ganache.
4. Add a Localhost network to MetaMask, and switch to it.
5. On the home screen, stop Ganache.
6. After about 10-20 seconds, you should see a "Still connecting" banner appear.
7. Switch to Ethereum Mainnet.
8. After about a second, the banner should go away.
9. Switch back to Localhost.
10. Wait for the banner to appear.
11. Start Ganache.
12. After about 10-20 seconds, the banner should go away.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

In this video, we have Ganache running and the Localhost network added. We stop Ganache, but a connection banner does not appear, and we have to switch to another screen and back to get it to do so. Then we start Ganache again, and the banner does not disappear — again, we have to switch away and the back to get it to do so.

https://github.com/user-attachments/assets/cee15501-b824-46dc-97bb-eb77c31dbd2e

### **After**

In this video, we have Ganache running and the Localhost network added. We stop Ganache, and a connection banner appears after 10-20 seconds. We start Ganache again, and the banner disappears after 10-20 seconds.

https://github.com/user-attachments/assets/5532d9bb-6f37-4c08-a915-b76f352539df

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.





<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Upgrades network-controller and related deps, migrates MetaMask middleware and pending/signature/tx flows to JsonRpcEngine v2 with InternalProvider, and updates tests/policies accordingly.
> 
> - **Middleware/Core (JsonRpcEngine v2 migration)**:
>   - Rewrite `app/scripts/lib/createMetamaskMiddleware` in TypeScript using `JsonRpcEngineV2`, returning `asLegacyMiddleware`.
>   - Add v2 middlewares `createPendingNonceMiddleware` and `createPendingTxMiddleware` in `app/scripts/lib/middleware/pending.ts` with new TS tests; remove JS versions.
>   - Update signature utils to use v2 `JsonRpcRequest` and `MiddlewareContext` (`app/scripts/lib/signature/util.ts`) and adjust tests.
>   - Update transaction utils to accept `JsonRpcRequest` and `MiddlewareContext`; plumb `requestContext` through `metamask-controller` and `addDappTransaction`.
> - **Provider type changes**:
>   - Replace `SafeEventEmitterProvider` with `InternalProvider` across code/tests (e.g., swaps tests, test provider stub, bridge page tests).
> - **Controller/init adjustments**:
>   - Remove legacy type-mismatch suppressions; minor messenger/registry access tweaks where needed.
> - **UI/Bridge**:
>   - Use `global.ethereumProvider` directly with new provider type.
> - **Policies/Configs/Tests**:
>   - Update LavaMoat policies for new package paths and provider engine.
>   - E2E metrics/state snapshots and helpers updated to handle `networksMetadata` shape and ordering.
>   - CODEOWNERS updated for TS middleware path.
> - **Dependencies**:
>   - Bump `@metamask/network-controller` to `^27.0.0`, `@metamask/eth-json-rpc-middleware` to `^22`, `@metamask/eth-json-rpc-provider` to `^6`, `@metamask/json-rpc-engine` to `^10.2.0`, and `@metamask/controller-utils` to `^11.16.0`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6e569773d9163f65e81aa9157b4e121f8d6baa40. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

rekmarks committed:
refactor: Bump eth-json-rpc-provider and related deps

rekmarks committed:
chore: lavamoat:auto

rekmarks committed:
refactor: Bump to migrated eth-json-rpc-middleware

metamaskbot committed:
Update LavaMoat policies

rekmarks committed:
chore: Lint

rekmarks committed:
chore: Add latest preview builds

rekmarks committed:
chore: Add latest preview builds

rekmarks committed:
refactor: Migrate wallet middleware hooks

rekmarks committed:
chore: Add latest preview builds

rekmarks committed:
chore: Add latest preview builds

rekmarks committed:
Merge branch 'main' into rekm/preview-ejrpcm-22.0

rekmarks committed:
refactor: Migrate MetamaskMiddleware to v2

rekmarks committed:
refactor: Migrate createMetamaskMiddleware to TypeScript

mcmire committed:
Bump network-controller to v26 (+ dependencies)

metamaskbot committed:
Update LavaMoat policies

mcmire committed:
Fix createTestProviderTools

mcmire committed:
Try fixing API specs test

metamaskbot committed:
Update LavaMoat policies

mcmire committed:
Merge remote-tracking branch 'origin/rekm/preview-ejrpcm-22.0' into bump-network-controller-to-v26

mcmire committed:
Merge branch 'main' into bump-network-controller-to-v26

metamaskbot committed:
Update LavaMoat policies

mcmire committed:
Fix e2e tests

mcmire committed:
Fix state logs e2e tests

mcmire committed:
Update preview builds

mcmire committed:
Merge branch 'main' into bump-network-controller-to-v26

metamaskbot committed:
Update LavaMoat policies

mcmire committed:
Fix lint

mcmire commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

cryptodev-2s committed:
Merge remote-tracking branch 'origin/main' into bump-network-controller-to-v26

cryptodev-2s committed:
Remove preview builds

cryptodev-2s commented:
@metamaskbot update-policies

cryptodev-2s committed:
Merge remote-tracking branch 'origin/main' into bump-network-controller-to-v26

metamaskbot committed:
Update LavaMoat policies

cryptodev-2s committed:
test: update E2E snapshots for NetworkOrderController changes

Update test snapshots to reflect the addition of multichain networks
(Bitcoin, Solana, Tron) and featured networks to the orderedNetworkList.

The NetworkOrderController now automatically includes 8 networks instead
of 3, causing snapshot mismatches in the 'errors before initialization'
test cases.

Updated snapshots:
- errors-before-init-opt-in-ui-state.json
- errors-before-init-opt-in-background-state.json

Gudahtt commented on test/e2e/tests/metrics/state-snapshots/errors-after-init-opt-in-ui-state.json:
```diff
@@ -202,12 +202,7 @@
     "networksWithTransactionActivity": "object",
     "selectedNetworkClientId": "string",
-    "networksMetadata": {
-      "networkConfigurationId": {
```
Is it intentional that we're hiding this information in Sentry snapshots now?

cryptodev-2s replied:
@Gudahtt I see now why this was removed because of this ci failure https://github.com/MetaMask/metamask-extension/actions/runs/19773997500/job/56664038305?pr=37966

mcmire replied:
See https://github.com/MetaMask/metamask-extension/pull/37966#discussion_r2577702136 for context.

@cryptodev-2s `networksMetadata` still shows up in state so I figured that we still wanted to represent it in some form in these snapshots. What do you thikn?

cryptodev-2s committed:
revert networksMetadata sentry state change

cryptodev-2s commented on test/e2e/tests/metrics/errors.spec.ts:
```diff
@@ -46,6 +46,7 @@ const maskedBackgroundFields = [
   'AppStateController.fullScreenGasPollTokens',
   'AppStateController.notificationGasPollTokens',
   'AppStateController.popupGasPollTokens',
+  'NetworkController.networksMetadata',
```
I don’t have the context on why this was excluded, so I reverted it for now because I couldn’t identify a reason to keep it excluded. https://github.com/MetaMask/metamask-extension/pull/37966/commits/bc70da10e59ce4714ba2bcc422f5b1d30d39cd94

Gudahtt replied:
Ah OK, makes sense. And this test does wait for initialization, and is post-onboarding, so I guess we can't stop these updates easily. Fair enough, we need to mask it then

cryptodev-2s committed:
revert last commit "revert networksMetadata sentry state change"

cryptodev-2s committed:
Merge remote-tracking branch 'origin/main' into bump-network-controller-to-v26

metamaskbot committed:
Update LavaMoat policies

cryptodev-2s committed:
test: clarify NetworkController.networksMetadata masking comment

Move networksMetadata to its own comment explaining it varies with
codebase changes, not runtime variability like poll tokens.

cryptodev-2s committed:
chore: improve NetworkController type mismatch comment

Clarify the type mismatch is due to NetworkController v27 adding
NetworkStatus.Degraded enum value.

cryptodev-2s committed:
Merge remote-tracking branch 'origin/main' into bump-network-controller-to-v26

cryptodev-2s commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

Gudahtt commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -81,8 +81,10 @@ export const TransactionControllerInit: ControllerInitFunction<
     getNetworkClientRegistry: (...args) =>
-      // @ts-expect-error - NetworkController registry type mismatch between peer dependencies
       networkController().getNetworkClientRegistry(...args),
+    // @ts-expect-error Type mismatch due to @metamask/network-controller version mismatch.
```
>So it seems that we need a way, when testing with preview builds, to just update a controller package at the root level and not deeper in the dependency tree. Would you agree?

Yes exactly. At least when testing breaking changes.

For previewing non-breaking changes, we'd want to use a resolution still, for any copies that match that major version.

mcmire committed:
Fix type errors

cryptodev-2s commented on app/scripts/controller-init/messengers/assets/token-rates-controller-messenger.ts:
```diff
@@ -13,9 +15,13 @@ import {
 
-type Actions = TokensControllerGetStateAction | NetworkControllerGetStateAction;
+type Actions =
+  | TokenRatesControllerActions
```
This error showed up after I merged main into your branch which did include this recent bump in assets-controllers https://github.com/MetaMask/metamask-extension/pull/38359/files

mcmire replied:
Hmm. What's strange is that `assets-controllers` v92 doesn't contain a bump to `network-controller` or any changes to the messengers. So I'm not really sure why this error is appearing now. But maybe it's another consequence of different versions of `network-controller` appearing in the dependency tree 🤔

mcmire replied:
It seems that `assets-controllers` was just upgraded to v93 which does include a bump to the new version of `network-controller` we are upgrading to in this PR, so this change doesn't seem to be necessary anymore: [ebfabbf](https://github.com/MetaMask/metamask-extension/pull/37966/commits/ebfabbf61a2c1808ad5f7a509c77a68dfaa0a9d5)

mcmire committed:
Merge branch 'main' into bump-network-controller-to-v26

mcmire committed:
Fix policies and lint errors

mcmire committed:
Fix policy

mcmire committed:
Bump node-forge to avoid audit

mcmire committed:
Add EOL

mcmire committed:
This change doesn't seem to be necessary anymore

mcmire commented on app/scripts/lib/middleware/pending.ts:
```diff
@@ -0,0 +1,78 @@
+    }
+    return await getPendingNonce(
+      address,
+      context.get('networkClientId') as string,
```
The middleware in `selected-network-controller` adds `networkClientId` to the request context so this is a safe assumption.

mcmire commented on app/scripts/lib/signature/util.ts:
```diff
@@ -20,28 +31,40 @@ async function handleSignature(
+    networkClientId: requestContext.get('networkClientId') as string,
+    securityAlertResponse: requestContext.get(
+      'securityAlertResponse',
+    ) as Record<string, Json>,
```
- `createOriginMiddleware` adds `origin` to the request context
- the middleware in `selected-network-controller` adds `networkClientId` to the request context
- `ppom-middleware` adds `securityAlertResponse` to the request context

It's true we are making assumptions that these are set, but I think these are safe assumptions.

mcmire committed:
Merge branch 'main' into bump-network-controller-to-v26

mcmire committed:
Fix policy

mcmire committed:
Fix policy

mcmire commented on lavamoat/browserify/main/policy.json:
Same as for `beta/policy.json`.

mcmire commented on lavamoat/webpack/mv2/policy.json:
I also don't see any surprising changes here. I've highlighted one change below, though.

mcmire commented on lavamoat/webpack/mv2/policy.json:
```diff
@@ -1441,6 +1446,7 @@
         "@metamask/rpc-errors": true,
         "@metamask/safe-event-emitter": true,
         "@metamask/utils": true,
+        "deep-freeze-strict": true,
```
This is a new dependency in `json-rpc-engine`, added in this PR: https://github.com/MetaMask/core/pull/6176

mcmire commented on lavamoat/webpack/mv3/policy.json:
I also don't see any surprising changes here. I've highlighted one change below, though.

mcmire committed:
Fix lint issues

mcmire commented on app/scripts/metamask-controller.js:
```diff
@@ -1100,14 +1100,17 @@ export default class MetamaskController extends EventEmitter {
       version,
       // account mgmt
-      getAccounts,
+      getAccounts: (requestOrigin) => getAccounts({ origin: requestOrigin }),
```
Pre-existing problem (if a problem).

Gudahtt commented on app/scripts/controller-init/token-list-controller-init.ts:
```diff
@@ -35,7 +35,6 @@ export const TokenListControllerInit: ControllerInitFunction<
   // TODO: Fix TokenListControllerMessenger type - add TokenListControllerActions & TokenListControllerEvents
   // TODO: Bump @metamask/network-controller to match assets-controllers
   const controller = new TokenListController({
-    // @ts-expect-error - Messenger type mismatch due to missing controller actions/events and dependency version mismatch
```
Hmm, I'm not seeing a change to `token-rates-controller-messenger.ts`

This error was caused by a NetworkController version mismatch, specifically due to the `NetworkController:getNetworkClientById` action that `TokenListController` uses

mcmire commented on app/scripts/controller-init/tokens-controller-init.ts:
```diff
@@ -19,10 +19,8 @@ export const TokensControllerInit: ControllerInitFunction<
   // TODO: Fix TokensControllerMessenger type - add TokensControllerActions & TokensControllerEvents
   // TODO: Bump @metamask/network-controller, @metamask/accounts-controller, @metamask/keyring-controller to match assets-controllers
   const controller = new TokensController({
-    // @ts-expect-error - Messenger type mismatch due to missing controller actions/events and dependency version mismatch
```
`assets-controllers` has been updated to v93 so the versions of `network-controller` should now match.

cryptodev-2s replied:
Yes, indeed. I noticed the recently merged PR updating `AssetsControllers`, which uses the latest version of `NetworkController` and no longer requires the change you made earlier: https://github.com/MetaMask/metamask-extension/pull/38521

Gudahtt commented on app/scripts/lib/signature/util.ts:
```diff
@@ -20,28 +31,40 @@ async function handleSignature(
+    method,
+    params: params as string[],
+    origin: requestContext.get('origin') as string,
+    networkClientId: requestContext.get('networkClientId') as string,
```
Ah I see, I didn't see this until after I left the comment on this in the `pending` middleware. Same situation as here I supposed.

Gudahtt commented on app/scripts/lib/transaction/util.test.ts:
```diff
@@ -66,12 +67,18 @@ const TRANSACTION_OPTIONS_MOCK: AddTransactionOptions = {
+
+const makeRequestContext = () =>
+  new MiddlewareContext<Record<PropertyKey, unknown>>({
+    origin: TRANSACTION_OPTIONS_MOCK.origin as string,
```
Nit: Unnecessary cast
```suggestion
    origin: TRANSACTION_OPTIONS_MOCK.origin,
```

Gudahtt commented on app/scripts/lib/transaction/util.ts:
```diff
@@ -67,17 +68,23 @@ export type AddTransactionRequest = FinalAddTransactionRequest & {
+  const actionId = String(id);
+
+  // TODO: Find a home for and define the appropriate MiddlewareContext type
+  const origin = requestContext.assertGet('origin') as string;
```
Isn't it just a matter of using the expected type for the context on line 72 above? I am not following why we need casts here either.

Gudahtt reviewed (approved):
LGTM!

I found a bunch of issues with the types, but it seems like at least some of these are known issues that are planned to be addressed later. I don't entirely follow why we've taken that approach, but as long as it gets addressed at some point that's fine.

adonesky1 reviewed (approved):
LGTM!

cortisiko reviewed (approved):
🌮 🌮

cryptodev-2s reviewed (approved):
LGTM!
